src/ecies.cpp

```diff
@@ -0,0 +1,444 @@
+#include "mococrw/ecies.h"
+#include "mococrw/symmetric_crypto.h"
+#include "mococrw/openssl_wrap.h"
+
+namespace mococrw {
+
+class ECIESImpl
+{
+public:
+    ECIESImpl(const std::shared_ptr<KeyDerivationFunction> kdf,
+              const size_t macKeySize,
+              std::function<std::unique_ptr<MessageAuthenticationCode> (const std::vector<uint8_t> &)> macFactoryFunc,
+              const size_t symCipherKeySize,
+              std::function<std::unique_ptr<SymmetricCipherI>(const std::vector<uint8_t>&,
+                                                              const SymmetricCipherOperation&)> symCipherFactoryFunc,
+              const std::vector<uint8_t> &S1,
+              const std::vector<uint8_t> &S2)
+        : kdf(kdf), _macKeySize(macKeySize), _macFactoryFunc(macFactoryFunc),
+          _symCipherKeySize(symCipherKeySize), _symCipherFactoryFunc(symCipherFactoryFunc),
+          _S1(S1), _S2(S2), _key_encrypt(symCipherKeySize), _key_mac(macKeySize) {}
+
+    virtual ~ECIESImpl()
+    {
+        /* Write zeros into _sharedSecret and both keys */
+        _key_encrypt.assign(_key_encrypt.size(), 0);
+        _key_mac.assign(_key_mac.size(), 0);
+        _sharedSecret.assign(_sharedSecret.size(), 0);
+    }
+
+protected:
+    void createSymmetricKeys()
+    {
+        std::vector<uint8_t> result = kdf->deriveKey(_sharedSecret, _macKeySize + _symCipherKeySize, _S1);
```

Comment:
_sharedSecret is only ever used in this function, and this function is always called after setting _sharedSecret.

Maybe _sharedSecret should not exist at all and be a parameter of this function? I don't see a good reason to keep it around after setting _key_encrypt and _key_mac.

The same thing actually applies for _key_encrypt and _key_mac, they are only used after calling this function to create the MAC and the symmetric encryption context.

Can we turn this into something similar to:
```c++
sharedSecret = ECDH(...);
auto keys = createSymmetricKeys(std::move(sharedSecret));
_mac = createMAC(std::move(keys.macKey));
_encryptor = createEncryptor(std::move(keys.encryptionKey));
```

This would also allow us to not care about secure deletion of the shared secret, mac key and encryption key from this class, because we could just delegate this to the MAC and symmetric crypto implementations if we never keep a copy of the key around.

Reply:
Done